File: day11_part2_2022.pl

use strict;
use warnings;
use List::Util qw(reduce);

sub parse {
    my $s = shift;
    my $m = {};
    my @lines = split("\n", $s);
    foreach my $item (split(", ", (split(": ", $lines[1]))[1])) {
        push @{$m->{items}}, int(toInt($item));
    }
    my @f = split(" ", (split("= ", $lines[2]))[1]);
    if ($f[1] eq "+") {
        if ($f[2] eq "old") {
            $m->{operation} = sub { return $_[0] + $_[0] };
        } else {
            $m->{operation} = sub { return $_[0] + toInt($f[2]) };
        }
    } elsif ($f[1] eq "*") {
        if ($f[2] eq "old") {
            $m->{operation} = sub { return $_[0] * $_[0] };
        } else {
            $m->{operation} = sub { return $_[0] * toInt($f[2]) };
        }
    }
    $lines[3] =~ /Test: divisible by (\d+)/;
    $m->{div} = $1;
    $lines[4] =~ /If true: throw to monkey (\d+)/;
    $m->{next}[0] = $1;
    $lines[5] =~ /If false: throw to monkey (\d+)/;
    $m->{next}[1] = $1;
    return $m;
}

sub monkeyBusiness {
    my ($monkeys, $rounds, $worry) = @_;
    my $div = reduce { $a * $b } 1, map { $_->{div} } @{$monkeys};

    for (my $i = 0; $i < $rounds; $i++) {
        foreach my $m (@{$monkeys}) {
            while (scalar @{$m->{items}} > 0) {
                $m->{inspections}++;
                my $item = $m->{operation}->($m->{items}[0]);
                if ($worry) {
                    $item %= $div;
                } else {
                    $item /= 3;
                }
                if ($item % $m->{div} == 0) {
                    push @{$monkeys->[$m->{next}[0]]{items}}, $item;
                } else {
                    push @{$monkeys->[$m->{next}[1]]{items}}, $item;
                }
                shift @{$m->{items}};
            }
        }
    }
    my @inspections = ();
    foreach my $m (@{$monkeys}) {
        push @inspections, $m->{inspections};
    }
    @inspections = sort { $b <=> $a } @inspections;
    return $inspections[0] * $inspections[1];
}

sub readAll {
    my $path = shift;
    open my $fh, '<', $path or die "Cannot open file: $!";
    local $/;
    my $content = <$fh>;
    close $fh;
    return $content;
}

sub toInt {
    my $s = shift;
    return int($s);
}

my @monkeys;
my $s = readAll("input.txt");
foreach my $m (split("\n\n", $s)) {
    push @monkeys, parse($m);
}

for (my $i = 0; $i < scalar @monkeys; $i++) {
    $monkeys[$i] = parse((split("\n\n", $s))[$i]);
}
print monkeyBusiness(\@monkeys, 10000, 1) . "\n";

File: julia/day11_part2_2022.jl

function parse_input(filename)
    monkeys = []
    open(filename, "r") do file
        while !eof(file)
            readline(file)  # Monkey header
            items_line = readline(file)
            op_line = readline(file)
            test_line = readline(file)
            true_line = readline(file)
            false_line = readline(file)
            readline(file)  # blank line or EOF

            # Parse starting items
            items = [parse(Int, x) for x in split(split(items_line, ": ")[2], ", ")]
            
            # Parse operation
            op_parts = split(split(op_line, "= ")[2])
            op_func = if op_parts[3] == "old"
                x -> x * x
            elseif op_parts[2] == "*"
                x -> x * parse(Int, op_parts[3])
            else
                x -> x + parse(Int, op_parts[3])
            end

            # Parse test conditions
            div_test = parse(Int, split(test_line)[end])
            true_monkey = parse(Int, split(true_line)[end]) + 1
            false_monkey = parse(Int, split(false_line)[end]) + 1

            push!(monkeys, Dict(
                "items" => items,
                "op" => op_func,
                "div_test" => div_test,
                "true_monkey" => true_monkey,
                "false_monkey" => false_monkey,
                "inspections" => 0
            ))
        end
    end
    return monkeys
end

function solve_part1(filename)
    monkeys = parse_input(filename)
    
    for _ in 1:20
        for monkey in monkeys
            while !isempty(monkey["items"])
                item = popfirst!(monkey["items"])
                monkey["inspections"] += 1
                
                # Inspect and reduce worry
                worry = floor(Int, monkey["op"](item) / 3)
                
                # Test and throw
                if worry % monkey["div_test"] == 0
                    push!(monkeys[monkey["true_monkey"]]["items"], worry)
                else
                    push!(monkeys[monkey["false_monkey"]]["items"], worry)
                end
            end
        end
    end
    
    # Calculate monkey business
    inspections = [m["inspections"] for m in monkeys]
    sort!(inspections, rev=true)
    return inspections[1] * inspections[2]
end

function solve_part2(filename)
    monkeys = parse_input(filename)
    
    # Calculate product of all divisibility tests
    mod_factor = prod(m["div_test"] for m in monkeys)
    
    for _ in 1:10000
        for monkey in monkeys
            while !isempty(monkey["items"])
                item = popfirst!(monkey["items"])
                monkey["inspections"] += 1
                
                # Inspect 
                worry = monkey["op"](item)
                
                # Keep worry manageable
                worry %= mod_factor
                
                # Test and throw
                if worry % monkey["div_test"] == 0
                    push!(monkeys[monkey["true_monkey"]]["items"], worry)
                else
                    push!(monkeys[monkey["false_monkey"]]["items"], worry)
                end
            end
        end
    end
    
    # Calculate monkey business
    inspections = [m["inspections"] for m in monkeys]
    sort!(inspections, rev=true)
    return inspections[1] * inspections[2]
end

# Read from input.txt and print result
function main()
    println("Part 1: ", solve_part1("input.txt"))
    println("Part 2: ", solve_part2("input.txt"))
end

main()
